core/src/nodes/listeners.rs

```diff
@@ -105,65 +125,93 @@ where
 {
     /// A new address is being listened on.
     NewAddress {
+        /// The listener that is listening on the new address.
+        listener_id: ListenerId,
         /// The new address that is being listened on.
         listen_addr: Multiaddr
     },
     /// An address is no longer being listened on.
     AddressExpired {
+        /// The listener that is no longer listening on the address.
+        listener_id: ListenerId,
         /// The new address that is being listened on.
         listen_addr: Multiaddr
     },
     /// A connection is incoming on one of the listeners.
     Incoming {
+        /// The listener that produced the upgrade.
+        listener_id: ListenerId,
         /// The produced upgrade.
         upgrade: TTrans::ListenerUpgrade,
         /// Address of the listener which received the connection.
         listen_addr: Multiaddr,
         /// Address used to send back data to the incoming client.
         send_back_addr: Multiaddr,
     },
-    /// A listener has closed, either gracefully or with an error.
+    /// A listener closed.
     Closed {
+        /// The ID of the listener that closed.
+        listener_id: ListenerId,
         /// The listener that closed.
         listener: TTrans::Listener,
-        /// The error that happened. `Ok` if gracefully closed.
-        result: Result<(), <TTrans::Listener as Stream>::Error>,
     },
+    /// A listener errored.
```

Comment:
That's too vague. It should mention that the listener can continue running, and that the error is informative.